### ISSUE TITLE
`dart pub cache clean` probably shouldn't delete globally activated packages

### ISSUE THREAD

**Issue by jamesderlin** [dart-lang/pub#3783]
Currently `dart pub cache clean` seems to do the equivalent of `rm -rf ~/.pub-cache`.  I claim that it instead should leave packages installed with `dart pub global activate` alone.

* `~/.pub-cache/bin` might be in the user's executable search path.  I doubt that most people would expect that `dart pub cache clean` would break some of their commands; expecting that requires users knowing which commands are Dart-based.
* I don't think that users can easily get back globally activated packages with some single command.  They'd have to remember which globally activated packages were installed.
* Doing the equivalent of `rm -rf ~/.pub-cache` doesn't offer much benefit anyway since people could easily do that themselves.  What *would* be useful is removing specific subdirectories within `.pub-cache` since users probably shouldn't be expected to understand `.pub-cache`'s the internal directory layout.

CC @sigurdm

**Comment by sigurdm** [comment]
The activated packages depend on the cached packages - so if we left them they would stop working anyways.

**Comment by jamesderlin** [comment]
Oh.  Hm.

1. Could the activated packages be compiled to standalone executables to remove dependencies on cached packages?
2. Or perhaps the wrapper scripts in `.pub-cache/bin` (or the `dart pub global run` command itself) could be made to automatically fetch missing dependencies?
3. Or the cheapest (but less-user-friendly) solution could be for `dart pub cache clean` to issue a warning, say something like:
    ```
    In addition to removing cached packages, the following commands in $HOME/.pub-cache/bin will be deleted:
      collect_coverage
      devtools
      format_coverage 
      test_with_coverage
      webdev
     ```
    It'd be even better if the warning also could display what command(s) to run to reinstall those packages, although that would be non-trivial for things like `coverage` that install multiple files to `.pub-cache/bin`.

**Comment by sigurdm** [comment]
> Could the activated packages be compiled to standalone executables to remove dependencies on cached packages?

Yes I think we will go in that direction! (Not sure of the exact design or timeline yet)

The biggest problem I see is that dart aot compilation removes support for dart:mirrors. But I think that might be worth it for the convenience of self-contained executables.
